keps/sig-node/3322-add-maxRestartTimes-to-podSpec/README.md

```diff
@@ -0,0 +1,981 @@
+like `Never`, then the Job API no longer needs to consider these logics anymore. Currently, it uses an annotation `batch.kubernetes.io/job-index-failure-count`
+to record the retires number, which looks not elegant.
+
+Pros:
+  * BackoffLimitPerIndex can reuse this functionality and no longer need to consider the restart times per index.
+  Specifically, it can avoid to use the annotation anymore, and works at a high level control by watching the pod status.
+
+Cons:
+  * If we support max restart times with `restartPolicy=Never`, it semantically means _we can never
+  restart the pod until we met the restart times. Then there's little difference with `restartPolicy=OnFailure`.
+
+#### Caveats
+
+- When a Job's template is set with restartPolicy=OnFailure, and `backoffLimit` is greater than `maxRestartTimesOnFailure`,
+then pod may fall into failure when it hits the `maxRestartTimesOnFailure`+1, Job will create a new pod rather than restore
+the original failed pod. This is not a break change because not breaking the API semantics but something we should notice.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- If `maxRestartTimesOnFailure` is configured in a recoverable circumstance with a small number retries.
+E.g. we set the `maxRestartTimesOnFailure=3`, and the network shakes with 4 times, the service will stuck
+in a failed state forever after 3 restarts, so end-users should be careful when configuring this field.
+However, this problem also exists in Job's backoffLimit, which is default to 6.
+- Even a well-written application can be restarted for several times in a long-running time, then we mark it as
+failed is somehow a little rude.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Add `maxRestartTimesOnFailure` to `pod.spec`:
+
+```golang
+type PodSpec struct {
+  // ...
+  RestartPolicy RestartPolicy
+  // maxRestartTimesOnFailure indicates the maximum count container can restart. It's the sum of
+  // all containers' restart times and only works when RestartPolicy is set to `OnFailure`.
+  // Default to nil which means will restart infinity.
+  // +optional
+  maxRestartTimesOnFailure *int
+}
+```
+
+- If Pod's `RestartPolicy` is `Always` or `Never`, `maxRestartTimesOnFailure` is default to nil, and will not apply.
+- If Pod's `RestartPolicy` is `OnFailure`, `maxRestartTimesOnFailure` is also default to nil, which means infinite
```

Comment:
two questions
1. if Pod's `RestartPolicy` is `OnFailure` and `maxRestartTimesOnFailure` is 0, invalid? or means `Never`?
2. is the `maxRestartTimesOnFailure` editable for the pod?

Reply:
IMO:

1) if `restartPolicy` is "OnFailure" and `maxRestarts` is 0, it is effectively "Never".  I don't think we need to special-case 0 to be a failure, but I don't feel strongly and could be argued either way.
2) let's start with "no" and see if there's really a need?

Reply:
Yeah, it feels like never to me too.